martha/martha_v3.js

```diff
@@ -40,38 +154,74 @@ function marthaV3Handler(req, res) {
         return;
     }
+    const parsedUrl = url.parse(dataObjectUri);
+    if (!parsedUrl.host || !parsedUrl.path) {
+        console.error(`"${url.format(parsedUrl)}" is missing a host and/or a path.`);
+        const failureResponse = new FailureResponse(BAD_REQUEST_ERROR_CODE, `"${dataObjectUri}" is not a properly-formatted URI.`);
+        res.status(BAD_REQUEST_ERROR_CODE).send(failureResponse);
+        return;
+    }
+    const {drsUrl, sendAuth, bondUrl, responseParser} = determineDrsType(parsedUrl);
+    console.log(`Converting DRS URI to HTTPS: ${dataObjectUri} -> ${drsUrl}`);
 
-    console.log(`Received URL '${dataObjectUri}' from IP '${req.ip}'`);
-    let dataObjectResolutionUrl;
+    let response;
     try {
-        dataObjectResolutionUrl = dataObjectUriToHttps(dataObjectUri);
+        if (sendAuth) {
+            response = await apiAdapter.getJsonFrom(drsUrl, auth);
+        } else {
+            response = await apiAdapter.getJsonFrom(drsUrl);
+        }
     } catch (err) {
+        console.log('Received error while resolving DRS URL.');
         console.error(err);
-        const failureResponse = new FailureResponse(BAD_REQUEST_ERROR_CODE, `The specified URL '${dataObjectUri}' is invalid`);
-        res.status(BAD_REQUEST_ERROR_CODE).send(failureResponse);
+
+        const errorStatusCode = err.status;
+        if (typeof errorStatusCode === 'undefined') {
+            const failureResponse = new FailureResponse(SERVER_ERROR_CODE, `Received error while resolving DRS URL. ${err.message}`);
+            res.status(SERVER_ERROR_CODE).send(failureResponse);
+        } else {
+            res.status(errorStatusCode).send(err);
+        }
         return;
     }
 
-    const bondProvider = determineBondProvider(dataObjectUri);
-    const dataObjectPromise = getDataObjectMetadata(dataObjectResolutionUrl, auth, bondProvider);
-    const bondPromise = maybeTalkToBond(req, bondProvider);
+    let drsResponse;
+    try {
+        drsResponse = responseParser(response);
+    } catch (err) {
+        console.log('Received error while parsing response from DRS URL.');
+        console.error(err);
+
+        const errorStatusCode = err.status;
+        if (typeof errorStatusCode === 'undefined') {
+            const failureResponse = new FailureResponse(SERVER_ERROR_CODE, `Received error while parsing response from DRS URL. ${err.message}`);
+            res.status(SERVER_ERROR_CODE).send(failureResponse);
+        } else {
+            res.status(errorStatusCode).send(err);
+        }
+        return;
+    }
 
-    return Promise.all([dataObjectPromise, bondPromise])
-        .then((rawResults) => {
-            res.status(200).send(aggregateResponses(rawResults));
-        })
-        .catch((err) => {
-            console.log('Received error while either contacting Bond or resolving drs url.');
+    let bondSA;
+    if (bondUrl && req && req.headers && req.headers.authorization) {
+        try {
```

Comment:
TOL: Would it be possible and make sense to combine all 3 `try.. catch` blocks in `marthaV3Handler` into single block `try... catch` block? Since most of catch portion is same I think it would be more readable if they are in the same catch block. I know that we would loose the specific error logs pin-pointing the exact location of error. To avoid that, the functions that are called inside the error block could catch these errors locally and add extra information before throwing back the error. So something like
```
let response, drsResponse, bondSA;
try {
  if (sendAuth) {
            response = await apiAdapter.getJsonFrom(drsUrl, auth);
        } else {
            response = await apiAdapter.getJsonFrom(drsUrl);
  }
 drsResponse = responseParser(response);
 if (bondUrl && req && req.headers && req.headers.authorization) {
   bondSA = await apiAdapter.getJsonFrom(bondUrl, auth);
 }
}
catch {
  <insert common error handling here>
}
```

Reply:
I did think about that, but I decided to land on the side of more specific error messages.  I'm open to changing it, though, although I'm not sure how to figure out which call resulted in the error since they're all calling the same function (`apiAdapter.getJsonFrom()`).

Reply:
We could separate out the portion that does drs resolving part and the part that talks to Bond into helper functions. So for example
```
function resolveDrsUri(sendAuth, auth, drsUrl) {
try {
  if (sendAuth) {
            response = await apiAdapter.getJsonFrom(drsUrl, auth);
        } else {
            response = await apiAdapter.getJsonFrom(drsUrl);
  }
 drsResponse = responseParser(response);
}
catch {
 <add additional error information>
}
}

function maybeTalkToBond(bondUrl, req) {
  try{
    if (bondUrl && req && req.headers && req.headers.authorization) {
     bondSA = await apiAdapter.getJsonFrom(bondUrl, auth);
    }
  }
  catch {
     <add additional error information>
  }
}

let response, drsResponse, bondSA;
try {
  drsResponse = resolveDrsUri(sendAuth, auth, drsUrl);
  bondSA = maybeTalkToBond(bondUrl, req);
}
catch {
  <insert common error handling here>
}
```
This again introduces 3 `try... catch` blocks (🤦‍♀️ ) but would make the main `marthaV3Handler` code more readable.
That being said I am not married to the idea of changing separating the 3 `try... catch` blocks. This was just TOL thought 😃

Reply:
Unless you feel strongly about it (or @kshakir weighs in on the other side), I'm content to leave it this way.